snuba_sdk/query_visitors.py

```diff
@@ -0,0 +1,321 @@
+        return ""
+
+    def _visit_limitby(self, limitby: Optional[LimitBy]) -> str:
+        if limitby is not None:
+            return f"LIMIT {self.translator.visit(limitby)}"
+        return ""
+
+    def _visit_limit(self, limit: Optional[Limit]) -> str:
+        if limit is not None:
+            return f"LIMIT {self.translator.visit(limit)}"
+        return ""
+
+    def _visit_offset(self, offset: Optional[Offset]) -> str:
+        if offset is not None:
+            return f"OFFSET {self.translator.visit(offset)}"
+        return ""
+
+    def _visit_granularity(self, granularity: Optional[Granularity]) -> str:
+        if granularity is not None:
+            return f"GRANULARITY {self.translator.visit(granularity)}"
+        return ""
+
+    def _visit_totals(self, totals: Optional[Totals]) -> str:
+        if totals is not None:
+            return f"TOTALS {self.translator.visit(totals)}"
+        return ""
+
+
+class Translator(Printer):
+    def __init__(self) -> None:
+        super().__init__(False)
+
+    def _combine(self, query: Query, returns: MutableMapping[str, str]) -> str:
+        formatted_query = super()._combine(query, returns)
+        body = {"dataset": query.dataset, "query": formatted_query}
+
+        return json.dumps(body)
+
+
+class Validator(QueryVisitor[None]):
+    def _combine(self, query: Query, returns: MutableMapping[str, None]) -> None:
+        # TODO: Contextual validations:
+        # - Must have certain conditions (project, timestamp, organization etc.)
+
+        if query.select is None:
+            raise InvalidQuery("query must have at least one column in select")
+
+        # - Orderby must be a field in select
+        if query.orderby is not None:
+            for o in query.orderby:
+                found = False
+                for s in query.select:
+                    if s == o.exp:
+                        found = True
+                        break
+
+                if not found:
+                    raise InvalidQuery(
+                        f"{o.exp} in orderby clause is missing from select clause"
+                    )
+
+        # - limit by must be a field in select
+        if query.limitby is not None:
+            found = False
+            for s in query.select:
+                if s == query.limitby.column:
+                    found = True
+                    break
+
+            if not found:
+                raise InvalidQuery(
+                    f"{query.limitby.column} in limitby clause is missing from select clause"
+                )
+
+        # Top level functions in the select clause must have an alias
+        non_aggregates = []
+        has_aggregates = False
+        for exp in query.select:
+            if isinstance(exp, Function) and not exp.alias:
+                raise InvalidQuery(f"{exp} must have an alias in the select")
+
+            if not isinstance(exp, Function) or not exp.is_aggregate():
```

Comment:
how does this identify aggregates inside a function?
`plus(count(event_id), count(trace_id))`